File: py/webhooks/src/figma_webhooks/sdk.py
"""
High-level SDK for Figma Webhooks API.
"""

from typing import List, Optional, Dict, Any, AsyncIterator

from .client import FigmaWebhooksClient
from .models import (
    Webhook,
    WebhookRequest,
    WebhookEvent,
    WebhookStatus,
    WebhookContext,
    CreateWebhookData,
    UpdateWebhookData,
    WebhooksResponse,
    WebhookRequestsResponse,
)
from .utils import (
    validate_webhook_endpoint,
    validate_passcode,
    validate_description,
    clean_webhook_data,
)


class FigmaWebhooksSDK:
    """High-level SDK for Figma Webhooks API."""
    
    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.figma.com",
        timeout: float = 30.0,
        max_retries: int = 3,
    ):
        """
        Initialize the Figma Webhooks SDK.
        
        Args:
            api_key: Figma API key
            base_url: Base URL for Figma API
            timeout: Request timeout in seconds
            max_retries: Maximum number of retries for failed requests
        """
        self.client = FigmaWebhooksClient(
            api_key=api_key,
            base_url=base_url,
            timeout=timeout,
            max_retries=max_retries,
        )
    
    async def __aenter__(self):
        """Async context manager entry."""
        await self.client.__aenter__()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.client.__aexit__(exc_type, exc_val, exc_tb)
    
    async def close(self) -> None:
        """Close the underlying client."""
        await self.client.close()
    
    # Webhook Management
    
    async def list_webhooks(
        self,
        *,
        context: Optional[WebhookContext] = None,
        context_id: Optional[str] = None,
        plan_api_id: Optional[str] = None,
        cursor: Optional[str] = None,
    ) -> WebhooksResponse:
        """
        List webhooks by context or plan.
        
        Args:
            context: Context type (TEAM, PROJECT, FILE)
            context_id: ID of the context
            plan_api_id: Plan API ID for getting all webhooks
            cursor: Pagination cursor
            
        Returns:
            Webhooks response with list of webhooks
        """
        params = {}
        if context:
            params["context"] = context.value
        if context_id:
            params["context_id"] = context_id
        if plan_api_id:
            params["plan_api_id"] = plan_api_id
        if cursor:
            params["cursor"] = cursor
        
        response = await self.client.get("/v2/webhooks", **params)
        return WebhooksResponse(**response)
    
    async def get_webhook(self, webhook_id: str) -> Webhook:
        """
        Get a webhook by ID.
        
        Args:
            webhook_id: ID of the webhook
            
        Returns:
            Webhook object
        """
        response = await self.client.get(f"/v2/webhooks/{webhook_id}")
        return Webhook(**response["webhook"])
    
    async def create_webhook(self, webhook_data: CreateWebhookData) -> Webhook:
        """
        Create a new webhook.
        
        Args:
            webhook_data: Webhook creation data
            
        Returns:
            Created webhook object
        """
        # Validate webhook data
        if not validate_webhook_endpoint(webhook_data.endpoint):
            raise ValueError("Invalid webhook endpoint URL")
        
        if not validate_passcode(webhook_data.passcode):
            raise ValueError("Invalid webhook passcode")
        
        if not validate_description(webhook_data.description):
            raise ValueError("Invalid webhook description")
        
        # Convert to dict and clean
        data = clean_webhook_data(webhook_data.model_dump(exclude_none=True))
        
        # Convert enums to values
        if "event_type" in data:
            data["event_type"] = data["event_type"].value if hasattr(data["event_type"], "value") else data["event_type"]
        if "context" in data:
            data["context"] = data["context"].value if hasattr(data["context"], "value") else data["context"]
        if "status" in data:
            data["status"] = data["status"].value if hasattr(data["status"], "value") else data["status"]
        
        response = await self.client.post("/v2/webhooks", json_data=data)
        return Webhook(**response["webhook"])
    
    async def update_webhook(self, webhook_id: str, webhook_data: UpdateWebhookData) -> Webhook:
        """
        Update an existing webhook.
        
        Args:
            webhook_id: ID of the webhook to update
            webhook_data: Webhook update data
            
        Returns:
            Updated webhook object
        """
        # Validate webhook data
        if webhook_data.endpoint and not validate_webhook_endpoint(webhook_data.endpoint):
            raise ValueError("Invalid webhook endpoint URL")
        
        if webhook_data.passcode and not validate_passcode(webhook_data.passcode):
            raise ValueError("Invalid webhook passcode")
        
        if webhook_data.description is not None and not validate_description(webhook_data.description):
            raise ValueError("Invalid webhook description")
        
        # Convert to dict and clean
        data = clean_webhook_data(webhook_data.model_dump(exclude_none=True))
        
        # Convert enums to values
        if "event_type" in data:
            data["event_type"] = data["event_type"].value if hasattr(data["event_type"], "value") else data["event_type"]
        if "status" in data:
            data["status"] = data["status"].value if hasattr(data["status"], "value") else data["status"]
        
        response = await self.client.put(f"/v2/webhooks/{webhook_id}", json_data=data)
        return Webhook(**response["webhook"])
    
    async def delete_webhook(self, webhook_id: str) -> bool:
        """
        Delete a webhook.
        
        Args:
            webhook_id: ID of the webhook to delete
            
        Returns:
            True if deletion was successful
        """
        await self.client.delete(f"/v2/webhooks/{webhook_id}")
        return True
    
    async def get_webhook_requests(self, webhook_id: str) -> WebhookRequestsResponse:
        """
        Get requests for a webhook.
        
        Args:
            webhook_id: ID of the webhook
            
        Returns:
            Webhook requests response
        """
        response = await self.client.get(f"/v2/webhooks/{webhook_id}/requests")
        return WebhookRequestsResponse(**response)
    
    # Convenience Methods
    
    async def create_file_webhook(
        self,
        file_id: str,
        endpoint: str,
        passcode: str,
        event_type: WebhookEvent = WebhookEvent.FILE_UPDATE,
        description: Optional[str] = None,
        status: WebhookStatus = WebhookStatus.ACTIVE,
    ) -> Webhook:
        """
        Create a webhook for a specific file.
        
        Args:
            file_id: Figma file ID
            endpoint: Webhook endpoint URL
            passcode: Webhook passcode
            event_type: Type of event to subscribe to
            description: Optional description
            status: Initial webhook status
            
        Returns:
            Created webhook object
        """
        webhook_data = CreateWebhookData(
            event_type=event_type,
            context=WebhookContext.FILE,
            context_id=file_id,
            endpoint=endpoint,
            passcode=passcode,
            description=description,
            status=status,
        )
        return await self.create_webhook(webhook_data)
    
    async def create_team_webhook(
        self,
        team_id: str,
        endpoint: str,
        passcode: str,
        event_type: WebhookEvent = WebhookEvent.FILE_UPDATE,
        description: Optional[str] = None,
        status: WebhookStatus = WebhookStatus.ACTIVE,
    ) -> Webhook:
        """
        Create a webhook for a team.
        
        Args:
            team_id: Figma team ID
            endpoint: Webhook endpoint URL
            passcode: Webhook passcode
            event_type: Type of event to subscribe to
            description: Optional description
            status: Initial webhook status
            
        Returns:
            Created webhook object
        """
        webhook_data = CreateWebhookData(
            event_type=event_type,
            context=WebhookContext.TEAM,
            context_id=team_id,
            endpoint=endpoint,
            passcode=passcode,
            description=description,
            status=status,
        )
        return await self.create_webhook(webhook_data)
    
    async def create_project_webhook(
        self,
        project_id: str,
        endpoint: str,
        passcode: str,
        event_type: WebhookEvent = WebhookEvent.FILE_UPDATE,
        description: Optional[str] = None,
        status: WebhookStatus = WebhookStatus.ACTIVE,
    ) -> Webhook:
        """
        Create a webhook for a project.
        
        Args:
            project_id: Figma project ID
            endpoint: Webhook endpoint URL
            passcode: Webhook passcode
            event_type: Type of event to subscribe to
            description: Optional description
            status: Initial webhook status
            
        Returns:
            Created webhook object
        """
        webhook_data = CreateWebhookData(
            event_type=event_type,
            context=WebhookContext.PROJECT,
            context_id=project_id,
            endpoint=endpoint,
            passcode=passcode,
            description=description,
            status=status,
        )
        return await self.create_webhook(webhook_data)
    
    async def pause_webhook(self, webhook_id: str) -> Webhook:
        """
        Pause a webhook.
        
        Args:
            webhook_id: ID of the webhook to pause
            
        Returns:
            Updated webhook object
        """
        webhook_data = UpdateWebhookData(status=WebhookStatus.PAUSED)
        return await self.update_webhook(webhook_id, webhook_data)
    
    async def activate_webhook(self, webhook_id: str) -> Webhook:
        """
        Activate a paused webhook.
        
        Args:
            webhook_id: ID of the webhook to activate
            
        Returns:
            Updated webhook object
        """
        webhook_data = UpdateWebhookData(status=WebhookStatus.ACTIVE)
        return await self.update_webhook(webhook_id, webhook_data)
    
    async def search_webhooks(
        self,
        *,
        event_type: Optional[WebhookEvent] = None,
        status: Optional[WebhookStatus] = None,
        context: Optional[WebhookContext] = None,
        plan_api_id: Optional[str] = None,
    ) -> List[Webhook]:
        """
        Search for webhooks with filters.
        
        Args:
            event_type: Filter by event type
            status: Filter by status
            context: Filter by context type
            plan_api_id: Plan API ID for searching across all contexts
            
        Returns:
            List of matching webhooks
        """
        webhooks = []
        
        # Get all webhooks (paginated if using plan_api_id)
        if plan_api_id:
            async for webhook_data in self.client.paginate("/v2/webhooks", plan_api_id=plan_api_id):
                webhook = Webhook(**webhook_data)
                
                # Apply filters
                if event_type and webhook.event_type != event_type:
                    continue
                if status and webhook.status != status:
                    continue
                if context and webhook.context != context:
                    continue
                
                webhooks.append(webhook)
        else:
            # Without plan_api_id, we can only search within specific contexts
            if not context:
                raise ValueError("Either plan_api_id or context must be provided")
            
            response = await self.list_webhooks(context=context)
            for webhook in response.webhooks:
                # Apply filters
                if event_type and webhook.event_type != event_type:
                    continue
                if status and webhook.status != status:
                    continue
                
                webhooks.append(webhook)
        
        return webhooks
    
    async def batch_get_webhooks(self, webhook_ids: List[str]) -> List[Optional[Webhook]]:
        """
        Get multiple webhooks by their IDs.
        
        Args:
            webhook_ids: List of webhook IDs
            
        Returns:
            List of webhook objects (None for webhooks that don't exist)
        """
        webhooks = []
        
        for webhook_id in webhook_ids:
            try:
                webhook = await self.get_webhook(webhook_id)
                webhooks.append(webhook)
            except Exception:
                webhooks.append(None)
        
        return webhooks

File: ts/webhooks/src/sdk.ts
/**
 * SDK facade for figma-webhooks
 * Provides ergonomic API for Figma Webhooks operations.
 *
 * Every Figma call routes through the generic client verbs
 * (`fetcher.get/post/put/delete`) against the v2 webhooks endpoints — the SDK
 * does NOT depend on bespoke `fetcher.createWebhook`/`paginateWebhooks`/… domain
 * methods (which the shipped `FigmaApiClient` does not implement). Pagination,
 * signature verification, and the event/context vocabularies live in this
 * service layer (R1), mirroring the Python edition's behavior.
 */

import { createHmac, timingSafeEqual } from "node:crypto";

/** Figma v2 webhook event types. */
export const WEBHOOK_EVENT_TYPES = [
  "PING",
  "FILE_UPDATE",
  "FILE_VERSION_UPDATE",
  "FILE_DELETE",
  "LIBRARY_PUBLISH",
  "FILE_COMMENT",
] as const;

/** Figma v2 webhook context types. */
export const WEBHOOK_CONTEXT_TYPES = ["team", "project", "file"] as const;

export interface FigmaWebhooksSDKConfig {
  /** FigmaApiClient instance (required) */
  fetcher: any;
  /** Custom logger */
  logger?: any;
}

/**
 * High-level SDK for Figma Webhooks API
 * Provides convenient methods for common webhook operations
 *
 * @example
 * import { FigmaApiClient } from '@figma-api/fetch';
 * import { FigmaWebhooksSDK } from 'figma-webhooks';
 *
 * const fetcher = new FigmaApiClient({ apiToken: process.env.FIGMA_TOKEN });
 * const sdk = new FigmaWebhooksSDK({ fetcher });
 */
export class FigmaWebhooksSDK {
  fetcher: any;
  logger: any;

  /**
   * Initialize the Figma Webhooks SDK
   * @param config - Configuration object
   * @param config.fetcher - FigmaApiClient instance (required)
   * @param config.logger - Custom logger
   */
  constructor(
    {
      fetcher,
      logger = console,
    }: FigmaWebhooksSDKConfig = {} as FigmaWebhooksSDKConfig,
  ) {
    if (!fetcher) {
      throw new Error(
        "fetcher parameter is required. Please create and pass a FigmaApiClient instance.",
      );
    }

    this.fetcher = fetcher;
    this.logger = logger;
  }

  // === Internal helpers (generic-verb plumbing) ===

  /**
   * Build the snake_case v2 create/update body Figma expects from a camelCase
   * options object, dropping undefined fields.
   */
  private _toWebhookBody(input: Record<string, any>): Record<string, any> {
    const map: Record<string, string> = {
      eventType: "event_type",
      contextId: "context_id",
      planApiId: "plan_api_id",
    };
    const body: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      if (value === undefined) continue;
      body[map[key] || key] = value;
    }
    return body;
  }

  /** Unwrap a `{ webhook }` envelope (Figma returns the created/updated webhook nested). */
  private _unwrapWebhook(response: any): any {
    return response?.webhook ?? response;
  }

  /** POST /v2/webhooks (create). */
  private async _createWebhook(input: Record<string, any>): Promise<any> {
    const response = await this.fetcher.post(
      "/v2/webhooks",
      this._toWebhookBody(input),
    );
    return this._unwrapWebhook(response);
  }

  /** Is this error a 404 from the shared error model? */
  private _isNotFound(error: any): boolean {
    return error?.code === "NOT_FOUND" || error?.meta?.status === 404;
  }

  // === Webhook Management Methods ===

  /**
   * Create a file update webhook
   * @returns Created webhook
   */
  async createFileWebhook({
    fileKey,
    endpoint,
    passcode,
    description,
    active = true,
  }: {
    fileKey: string;
    endpoint: string;
    passcode: string;
    description?: string;
    active?: boolean;
  }): Promise<any> {
    return this._createWebhook({
      eventType: "FILE_UPDATE",
      context: "file",
      contextId: fileKey,
      endpoint,
      passcode,
      status: active ? "ACTIVE" : "PAUSED",
      description,
    });
  }

  /**
   * Create a project webhook for all files in a project
   * @returns Created webhook
   */
  async createProjectWebhook({
    projectId,
    eventType = "FILE_UPDATE",
    endpoint,
    passcode,
    description,
    active = true,
  }: {
    projectId: string;
    eventType?: string;
    endpoint: string;
    passcode: string;
    description?: string;
    active?: boolean;
  }): Promise<any> {
    return this._createWebhook({
      eventType,
      context: "project",
      contextId: projectId,
      endpoint,
      passcode,
      status: active ? "ACTIVE" : "PAUSED",
      description,
    });
  }

  /**
   * Create a team webhook for all team activity
   * @returns Created webhook
   */
  async createTeamWebhook({
    teamId,
    eventType = "FILE_UPDATE",
    endpoint,
    passcode,
    description,
    active = true,
  }: {
    teamId: string;
    eventType?: string;
    endpoint: string;
    passcode: string;
    description?: string;
    active?: boolean;
  }): Promise<any> {
    return this._createWebhook({
      eventType,
      context: "team",
      contextId: teamId,
      endpoint,
      passcode,
      status: active ? "ACTIVE" : "PAUSED",
      description,
    });
  }

  /**
   * Get webhook by ID with enhanced error handling
   * @param webhookId - Webhook ID
   * @returns Webhook data or null if not found
   */
  async getWebhook(webhookId: string): Promise<any | null> {
    try {
      const response = await this.fetcher.get(`/v2/webhooks/${webhookId}`);
      return this._unwrapWebhook(response);
    } catch (error: any) {
      if (this._isNotFound(error)) {
        return null;
      }
      throw error;
    }
  }

  /**
   * List all webhooks for a context
   * @param options - Query options ({ context, contextId, planApiId })
   * @returns Array of webhooks
   */
  async listWebhooks(options: Record<string, any> = {}): Promise<any[]> {
    const response = await this.fetcher.get(
      "/v2/webhooks",
      this._toWebhookBody(options),
    );
    return response.webhooks || [];
  }

  /**
   * List all webhooks across all accessible contexts, following pagination.
   * @param planApiId - Plan API ID
   * @returns Array of all webhooks
   */
  async listAllWebhooks(planApiId: string): Promise<any[]> {
    const webhooks: any[] = [];
    let cursor: string | undefined;

    do {
      const params: Record<string, any> = { plan_api_id: planApiId };
      if (cursor) params.cursor = cursor;
      const response = await this.fetcher.get("/v2/webhooks", params);
      if (Array.isArray(response.webhooks)) {
        webhooks.push(...response.webhooks);
      }
      cursor = response.next_page || undefined;
    } while (cursor);

    return webhooks;
  }

  /**
   * Update webhook with partial data
   * @param webhookId - Webhook ID
   * @param updates - Fields to update
   * @returns Updated webhook
   */
  async updateWebhook(
    webhookId: string,
    updates: Record<string, any>,
  ): Promise<any> {
    const response = await this.fetcher.put(
      `/v2/webhooks/${webhookId}`,
      this._toWebhookBody(updates),
    );
    return this._unwrapWebhook(response);
  }

  /**
   * Delete webhook with confirmation
   * @param webhookId - Webhook ID
   * @returns True if successfully deleted
   */
  async deleteWebhook(webhookId: string): Promise<boolean> {
    try {
      await this.fetcher.delete(`/v2/webhooks/${webhookId}`);
      return true;
    } catch (error: any) {
      if (this._isNotFound(error)) {
        return false; // Already deleted
      }
      throw error;
    }
  }

  /**
   * Pause webhook
   * @param webhookId - Webhook ID
   * @returns Updated webhook
   */
  async pauseWebhook(webhookId: string): Promise<any> {
    return this.updateWebhook(webhookId, { status: "PAUSED" });
  }

  /**
   * Activate webhook
   * @param webhookId - Webhook ID
   * @returns Updated webhook
   */
  async activateWebhook(webhookId: string): Promise<any> {
    return this.updateWebhook(webhookId, { status: "ACTIVE" });
  }

  // === Webhook Monitoring Methods ===

  /**
   * Get webhook delivery history for debugging
   * @param webhookId - Webhook ID
   * @returns Recent webhook requests
   */
  async getWebhookHistory(webhookId: string): Promise<any[]> {
    const response = await this.fetcher.get(
      `/v2/webhooks/${webhookId}/requests`,
    );
    return response.requests || [];
  }

  /**
   * Check webhook health by analyzing recent deliveries
   * @param webhookId - Webhook ID
   * @returns Health status report
   */
  async checkWebhookHealth(webhookId: string): Promise<any> {
    const requests = await this.getWebhookHistory(webhookId);

    if (requests.length === 0) {
      return {
        status: "unknown",
        message: "No recent webhook deliveries found",
        successRate: null,
        lastDelivery: null,
      };
    }

    const successful = requests.filter(
      (req: any) =>
        req.response_info && parseInt(req.response_info.status, 10) < 400,
    );

    const successRate = successful.length / requests.length;
    const lastRequest = requests[0]; // Most recent

    let status = "healthy";
    let message = `Webhook is healthy (${Math.round(successRate * 100)}% success rate)`;

    if (successRate < 0.8) {
      status = "degraded";
      message = `Webhook is experiencing issues (${Math.round(successRate * 100)}% success rate)`;
    }

    if (successRate === 0) {
      status = "failing";
      message = "Webhook is failing all deliveries";
    }

    return {
      status,
      message,
      successRate,
      lastDelivery: lastRequest?.request_info?.sent_at,
      totalRequests: requests.length,
      successfulRequests: successful.length,
      failedRequests: requests.length - successful.length,
    };
  }

  // === Bulk Operations ===

  /**
   * Create multiple webhooks with the same configuration
   * @param contextIds - Array of context IDs
   * @param webhookConfig - Base webhook configuration
   * @returns Results with created webhooks and errors
   */
  async createBulkWebhooks(
    contextIds: string[],
    webhookConfig: Record<string, any>,
  ): Promise<any> {
    const results: { created: any[]; errors: any[] } = {
      created: [],
      errors: [],
    };

    for (const contextId of contextIds) {
      try {
        const webhook = await this._createWebhook({
          ...webhookConfig,
          contextId,
        });
        results.created.push(webhook);
      } catch (error: any) {
        results.errors.push({
          contextId,
          error: error.message,
          code: error.code,
        });
      }
    }

    return results;
  }

  /**
   * Delete multiple webhooks
   * @param webhookIds - Array of webhook IDs
   * @returns Results with deletion status
   */
  async deleteBulkWebhooks(webhookIds: string[]): Promise<any> {
    const results: { deleted: any[]; errors: any[] } = {
      deleted: [],
      errors: [],
    };

    for (const webhookId of webhookIds) {
      try {
        await this.deleteWebhook(webhookId);
        results.deleted.push(webhookId);
      } catch (error: any) {
        results.errors.push({
          webhookId,
          error: error.message,
          code: error.code,
        });
      }
    }

    return results;
  }

  /**
   * Pause multiple webhooks
   * @param webhookIds - Array of webhook IDs
   * @returns Results with pause status
   */
  async pauseBulkWebhooks(webhookIds: string[]): Promise<any> {
    const results: { paused: any[]; errors: any[] } = {
      paused: [],
      errors: [],
    };

    for (const webhookId of webhookIds) {
      try {
        const webhook = await this.pauseWebhook(webhookId);
        results.paused.push(webhook);
      } catch (error: any) {
        results.errors.push({
          webhookId,
          error: error.message,
          code: error.code,
        });
      }
    }

    return results;
  }

  // === Webhook Discovery ===

  /**
   * Find webhooks by endpoint URL
   * @param endpoint - Endpoint URL to search for
   * @param planApiId - Plan API ID
   * @returns Matching webhooks
   */
  async findWebhooksByEndpoint(
    endpoint: string,
    planApiId: string,
  ): Promise<any[]> {
    const allWebhooks = await this.listAllWebhooks(planApiId);
    return allWebhooks.filter((webhook: any) => webhook.endpoint === endpoint);
  }

  /**
   * Find webhooks by event type
   * @param eventType - Event type to search for
   * @param planApiId - Plan API ID
   * @returns Matching webhooks
   */
  async findWebhooksByEventType(
    eventType: string,
    planApiId: string,
  ): Promise<any[]> {
    const allWebhooks = await this.listAllWebhooks(planApiId);
    return allWebhooks.filter(
      (webhook: any) => webhook.event_type === eventType,
    );
  }

  /**
   * Find inactive webhooks
   * @param planApiId - Plan API ID
   * @returns Inactive webhooks
   */
  async findInactiveWebhooks(planApiId: string): Promise<any[]> {
    const allWebhooks = await this.listAllWebhooks(planApiId);
    return allWebhooks.filter((webhook: any) => webhook.status === "PAUSED");
  }

  // === Convenience Methods ===

  /**
   * Test webhook endpoint connectivity
   * @param endpoint - Endpoint URL to test
   * @returns Test results
   */
  async testWebhookEndpoint(endpoint: string): Promise<any> {
    const testPayload = {
      event_type: "PING",
      timestamp: new Date().toISOString(),
      webhook_id: "test",
    };

    try {
      // Use the global fetch for arbitrary (non-Figma) endpoints.
      const response: any = await globalThis.fetch(endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-Figma-Event": "PING",
        },
        body: JSON.stringify(testPayload),
      });

      return {
        reachable: true,
        status: response.status,
        statusText: response.statusText,
        headers: Object.fromEntries(response.headers.entries()),
      };
    } catch (error: any) {
      return {
        reachable: false,
        error: error.message,
      };
    }
  }

  /**
   * Verify a webhook payload signature.
   *
   * Figma signs the request body with an HMAC-SHA256 keyed by the webhook
   * passcode and delivers it in the `X-Figma-Signature` header. This compares
   * the computed digest to the provided signature in constant time.
   *
   * @param payload - Raw webhook payload (the exact request body bytes/string)
   * @param signature - The `X-Figma-Signature` header value (hex)
   * @param passcode - The webhook passcode used as the HMAC key
   * @returns Whether the signature is valid
   */
  verifySignature(
    payload: string,
    signature: string,
    passcode: string,
  ): boolean {
    if (!payload || !signature || !passcode) return false;
    const expected = createHmac("sha256", passcode)
      .update(payload)
      .digest("hex");
    const a = Buffer.from(expected);
    const b = Buffer.from(signature);
    if (a.length !== b.length) return false;
    return timingSafeEqual(a, b);
  }

  /**
   * Get supported event types
   * @returns Available event types
   */
  getSupportedEventTypes(): string[] {
    return [...WEBHOOK_EVENT_TYPES];
  }

  /**
   * Get supported context types
   * @returns Available context types
   */
  getSupportedContextTypes(): string[] {
    return [...WEBHOOK_CONTEXT_TYPES];
  }

  /**
   * Clean up resources
   */
  async close(): Promise<void> {
    // Clean up any resources if needed
    this.logger.debug("FigmaWebhooksSDK closed");
  }
}

export default FigmaWebhooksSDK;
